spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/Graph.scala

```diff
@@ -212,6 +216,17 @@ abstract class Graph[T: ClassTag](
         || inputNames.contains(n.element.getName()))
     backwardTargets.foreach(_ -> dummyBackwardEnd)
     backwardGraph = dummyBackwardEnd.graph(true)
+
+    // Check if gradInput is empty for each input
+    isGradInputAvailable = inputs.map(_ => false).toArray
+    backwardGraph.DFS.foreach(curNode => {
+      inputs.zipWithIndex.map { case (n, i) =>
+        if (curNode.element.getName() == n.element.getName() && !isStopGradient(n.element)) {
+          isGradInputAvailable(i) = true
```

Comment:
This looks expensive. In addition, one can stop gradient in the middle (instead of on the input node) to make the gradInput of the input node empty, no?

Reply:
It happens in graph build time.

If one stop gradient in the middle, which makes gradient cannot be back propagated to some inputs, these inputs won be in the backwardGraph. So the gradInput of such inputs are empty.